streaming/src/test/scala/org/apache/spark/streaming/util/WriteAheadLogSuite.scala

```diff
@@ -190,283 +281,150 @@ class WriteAheadLogSuite extends SparkFunSuite with BeforeAndAfter {
-      }
-    } catch {
-      case ex: EOFException =>
-    } finally {
-      reader.close()
+  test("BatchedWriteAheadLog - name log with aggregated entries with the timestamp of last entry") {
+    val wal = new BatchedWriteAheadLog(fileBasedWAL, sparkConf)
+    // block the write so that we can batch some records
+    val promise = writeBlockingPromise(fileBasedWAL)
+
+    val event1 = "hello"
+    val event2 = "world"
+    val event3 = "this"
+    val event4 = "is"
+    val event5 = "doge"
+
+    eventFuture(wal, event1, 3L) // 3 will automatically be flushed for the first write
+    // rest of the records will be batched while it takes 3 to get written
+    eventFuture(wal, event2, 5L)
+    eventFuture(wal, event3, 8L)
+    eventFuture(wal, event4, 12L)
+    eventFuture(wal, event5, 10L)
+    promise.success(true)
+
+    eventually(Eventually.timeout(1 second)) {
+      verify(fileBasedWAL, times(1)).write(any[ByteBuffer], meq(3L))
+      // the file name should be the timestamp of the last record, as events should be naturally
+      // in order of timestamp, and we need the last element.
+      verify(fileBasedWAL, times(1)).write(any[ByteBuffer], meq(10L))
     }
-    buffer
-  }
-
-  /** Read all the data from a log file using reader class and return the list of byte buffers. */
-  def readDataUsingReader(file: String): Seq[String] = {
-    val reader = new FileBasedWriteAheadLogReader(file, hadoopConf)
-    val readData = reader.toList.map(byteBufferToString)
-    reader.close()
-    readData
   }
 
-  /** Read all the data in the log file in a directory using the WriteAheadLog class. */
-  def readDataUsingWriteAheadLog(logDirectory: String): Seq[String] = {
-    val wal = new FileBasedWriteAheadLog(new SparkConf(), logDirectory, hadoopConf, 1, 1,
-      closeFileAfterWrite = false)
-    val data = wal.readAll().asScala.map(byteBufferToString).toSeq
+  test("BatchedWriteAheadLog - shutdown properly") {
+    val wal = new BatchedWriteAheadLog(fileBasedWAL, sparkConf)
     wal.close()
-    data
-  }
-
-  /** Get the log files in a direction */
-  def getLogFilesInDirectory(directory: String): Seq[String] = {
-    val logDirectoryPath = new Path(directory)
-    val fileSystem = HdfsUtils.getFileSystemForPath(logDirectoryPath, hadoopConf)
-
-    if (fileSystem.exists(logDirectoryPath) && fileSystem.getFileStatus(logDirectoryPath).isDir) {
-      fileSystem.listStatus(logDirectoryPath).map { _.getPath() }.sortBy {
-        _.getName().split("-")(1).toLong
-      }.map {
-        _.toString.stripPrefix("file:")
-      }
-    } else {
-      Seq.empty
-    }
-  }
-
-  def generateRandomData(): Seq[String] = {
-    (1 to 100).map { _.toString }
-  }
-
-  implicit def stringToByteBuffer(str: String): ByteBuffer = {
-    ByteBuffer.wrap(Utils.serialize(str))
-  }
-
-  implicit def byteBufferToString(byteBuffer: ByteBuffer): String = {
-    Utils.deserialize[String](byteBuffer.array)
+    verify(fileBasedWAL, times(1)).close()
```

Comment:
I think this unit test should also test that if the batchWAL is closed, then outstanding write requests will complete with error, and not get blocked indefinitely. Very crucial behavior to avoid hard-to-debug deadlocks.